core/src/main/scala/org/apache/spark/rdd/BlockRDD.scala

```diff
@@ -66,7 +67,9 @@ class BlockRDD[T: ClassTag](@transient sc: SparkContext, @transient val blockIds
     blockIds.foreach { blockId =>
       sc.env.blockManager.master.removeBlock(blockId)
     }
-    _isValid = false
+    if (_setInvalid) {
```

Comment:
So rather than having this maybe the subclass of this could override `isValid`. It's a bit confusing as it stands.